### PR TITLE
Implement make_pressure_rhs_compatible for compressible computations

### PR DESCRIPTION
**Pull request by tjhei**
Completely re-implement make_pressure_rhs_compatible to handle all
combinations of
1. CG vs DG
2. direct vs iterative solver
3. melt vs no melt

(except melt and DG)

and add documentation and tests.

Bonus: remove wrong assert in denormalize_pressure that would trigger for parallel DG computations with a direct solver.

### CODE REVIEW

**Comment by jdannberg** [comment]
This doesn't really have to do anything with your changes, but while looking at this I wondered if we should issue a warning if the correction of the right-hand side is big (or at least write out the value of the correction somewhere). The feature is intended to make a small correction when for example the discretization of the velocity boundary conditions causes in- and outflow to be not exactly balanced. But there's nothing to prevent this functionality to also make a big correction if there is a big mass in- or outflux because someone has chosen boundary conditions in a way that in- or outflow are very different.

File: source/simulator/helper_functions.cc [hunk 1]
```diff
@@ -977,39 +975,75 @@ namespace aspect
...
-    // In the following we integrate the right hand side. This integral is the
-    // correction term that needs to be added to the pressure right hand side.
-    // (so that the integral of right hand side is set to zero).
-    if (!parameters.include_melt_transport && introspection.block_indices.velocities != introspection.block_indices.pressure)
+    if ((!parameters.use_locally_conservative_discretization)
+        &&
+        (!parameters.include_melt_transport)
+        &&
+        (introspection.block_indices.velocities != introspection.block_indices.pressure))
       {
```
Comment (jdannberg) [inline source/simulator/helper_functions.cc]: Does is make sense to also add a comment here that this is the 'normal' case (FE_Q, no melt, iterative solver)?

File: source/simulator/helper_functions.cc [hunk 2]
```diff
@@ -1019,6 +1053,64 @@ namespace aspect
 
         vector.compress(VectorOperation::add);
       }
+    else
+      {
+        // Locally conservative with or without direct solver and with or
+        // without melt: grab a pickaxe and do everything by hand!
```
Comment (jdannberg) [inline source/simulator/helper_functions.cc]: I guess I am not really the right person to look at this part, maybe @bangerth can have a look...
But did you also do a test with melt and DG? Or is that something that doesn't work?

File: source/simulator/helper_functions.cc (hunk 1 above)
Comment (tjhei) [inline source/simulator/helper_functions.cc]: there is a long comment 2 lines below.

File: source/simulator/helper_functions.cc (hunk 2 above)
Comment (tjhei) [inline source/simulator/helper_functions.cc]: I have not tested melt+DG. Suggestions for a test?

**Comment by tjhei** [comment]
Oh, of course melt+DG doesn't work for other reasons. Good to go.

**Comment by gassmoeller** [comment]
Looks correct to me. Thanks! Looking forward to running compressible DG-pressure models :smile:

**Pull request merged**